applications/GeoMechanicsApplication/custom_utilities/transport_equation_utilities.hpp

```diff
@@ -79,6 +86,174 @@ class GeoTransportEquationUtilities
+
+        // create general parameters of retention law
+        RetentionLaw::Parameters RetentionParameters(rProp, rCurrentProcessInfo);
+
+        Matrix mass_matrix_contribution = ZeroMatrix(block_element_size, block_element_size);
+
+        // Defining shape functions and the determinant of the jacobian at all integration points
+        // Loop over integration points
+        Matrix Nu                 = ZeroMatrix(dimension, number_U_nodes * dimension);
+        Matrix aux_density_matrix = ZeroMatrix(dimension, number_U_nodes * dimension);
+        Matrix density_matrix     = ZeroMatrix(dimension, dimension);
+
+        for (unsigned int g_point = 0; g_point < integration_points.size(); ++g_point) {
+            const double degree_of_saturation = CalculateDegreeOfSaturation(
+                g_point, np_container, pressure_vector, RetentionParameters, rRetentionLawVector);
+
+            const double density = CalculateSoilDensity(degree_of_saturation, rProp);
+
+            GeoElementUtilities::AssembleDensityMatrix(density_matrix, density);
+
+            // Setting the shape function matrix
+            SizeType index = 0;
+            for (SizeType i = 0; i < number_U_nodes; ++i) {
+                for (SizeType i_dim = 0; i_dim < dimension; ++i_dim) {
+                    Nu(i_dim, index) = nu_container(g_point, i);
+                    index++;
+                }
+            }
+            noalias(aux_density_matrix) = prod(density_matrix, Nu);
+
+            // Adding contribution to Mass matrix
+            const double integration_coefficient_initial_configuration =
+                CalculateIntegrationCoefficientInitialConfiguration(
+                    g_point, rGeom, integration_points, IntegrationMethod, rStressStatePolicy);
+            noalias(mass_matrix_contribution) +=
+                prod(trans(Nu), aux_density_matrix) * integration_coefficient_initial_configuration;
+        }
+
+        // Distribute mass block matrix into the elemental matrix
+        const SizeType element_size = block_element_size + number_P_nodes;
+        Matrix         mass_matrix  = ZeroMatrix(element_size, element_size);
+
+        for (SizeType i = 0; i < number_U_nodes * dimension; ++i) {
+            for (SizeType j = 0; j < number_U_nodes * dimension; ++j) {
+                mass_matrix(i, j) += mass_matrix_contribution(i, j);
+            }
+        }
+        return mass_matrix;
+    }
+
+    static double CalculateFluidPressure(const Vector& rNp, const Vector& rPressureVector)
+    {
+        return inner_prod(rNp, rPressureVector);
+    }
+
+    static Vector GetSolutionVector(SizeType NumberPNodes, const Geometry<Node>& rGeom, const Variable<double>& SolutionVariable)
+    {
+        Vector solution_vector(NumberPNodes);
+        for (SizeType i = 0; i < NumberPNodes; ++i) {
+            solution_vector[i] = rGeom[i].FastGetSolutionStepValue(SolutionVariable);
+        }
```

Comment:
This can be a transform 😃 :
```suggestion
        std::transform(rGeom.begin(), rGeom.end(), solution_vector.begin(),
                       [&SolutionVariable](const auto& node) {
            return node.FastGetSolutionStepValue(SolutionVariable);
        });
```

Reply:
Great! Used it.